File: truevfs-ext/truevfs-ext-pacemaker/src/main/scala/net/java/truevfs/ext/pacemaker/AspectController.scala
/*
 * Copyright (C) 2005-2015 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.ext.pacemaker

import java.io.{InputStream, OutputStream}
import java.nio.channels.SeekableByteChannel

import javax.annotation._
import net.java.truecommons.cio.Entry._
import net.java.truecommons.cio._
import net.java.truecommons.shed._
import net.java.truevfs.kernel.spec._

/** Calls a template method to apply an aspect to every file system operation.
  *
  * @see #apply
  * @author Christian Schlichtherle
  */
private abstract class AspectController(controller: FsController) extends FsDecoratingController(controller) {

  /**
    * Applies the aspect to the given file system operation.
    *
    * @param  operation the file system operation to apply an aspect to.
    * @return The return value of the file system operation.
    */
  protected def apply[V](operation: () => V): V

  override def node(options: AccessOptions, name: FsNodeName): FsNode = {
    apply(() => controller node(options, name))
  }

  override def checkAccess(options: AccessOptions, name: FsNodeName, types: BitField[Access]): Unit = {
    apply(() => controller checkAccess(options, name, types))
  }

  override def setReadOnly(options: AccessOptions, name: FsNodeName): Unit = {
    apply(() => controller setReadOnly(options, name))
  }

  override def setTime(options: AccessOptions, name: FsNodeName, times: java.util.Map[Access, java.lang.Long]): Boolean = {
    apply(() => controller setTime(options, name, times))
  }

  override def setTime(options: AccessOptions, name: FsNodeName, types: BitField[Access], value: Long): Boolean = {
    apply(() => controller setTime(options, name, types, value))
  }

  override def input(options: AccessOptions, name: FsNodeName): InputSocket[_ <: Entry] = {
    new Input(controller input(options, name)): InputSocket[_ <: Entry]
  }

  private class Input(socket: InputSocket[_ <: Entry]) extends AbstractInputSocket[Entry] {

    override def target(): Entry = apply(() => socket.target())

    override def stream(peer: OutputSocket[_ <: Entry]): InputStream = apply(() => socket stream peer)

    override def channel(peer: OutputSocket[_ <: Entry]): SeekableByteChannel = apply(() => socket channel peer)
  }

  override def output(options: AccessOptions, name: FsNodeName, @CheckForNull template: Entry): OutputSocket[_ <: Entry] = {
    new Output(controller output(options, name, template)): OutputSocket[_ <: Entry]
  }

  private class Output(socket: OutputSocket[_ <: Entry]) extends AbstractOutputSocket[Entry] {

    override def target(): Entry = apply(() => socket.target())

    override def stream(peer: InputSocket[_ <: Entry]): OutputStream = apply(() => socket stream peer)

    override def channel(peer: InputSocket[_ <: Entry]): SeekableByteChannel = apply(() => socket channel peer)
  }

  override def make(options: AccessOptions, name: FsNodeName, tµpe: Type, @CheckForNull template: Entry): Unit = {
    apply(() => controller make(options, name, tµpe, template))
  }

  override def unlink(options: AccessOptions, name: FsNodeName): Unit = {
    apply(() => controller unlink(options, name))
  }

  override def sync(options: BitField[FsSyncOption]): Unit = {
    apply(() => controller sync options)
  }
}

File: truevfs-ext/truevfs-ext-pacemaker/src/main/java/net/java/truevfs/ext/pacemaker/AspectController.java
/*
 * Copyright (C) 2005-2020 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.ext.pacemaker;

import net.java.truecommons.cio.*;
import net.java.truecommons.shed.BitField;
import net.java.truevfs.kernel.spec.*;

import javax.annotation.CheckForNull;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.channels.SeekableByteChannel;
import java.util.Map;

/**
 * Calls a template method to apply an aspect to all file system operations except {@link #sync(BitField)}.
 *
 * @see #apply
 * @author Christian Schlichtherle
 */
abstract class AspectController extends FsDecoratingController {

    AspectController(FsController controller) {
        super(controller);
    }

    /**
     * Applies the aspect to the given file system operation.
     *
     * @param  op the file system operation to apply an aspect to.
     * @return The return value of the file system operation.
     */
    abstract <T> T apply(Op<T> op) throws IOException;

    interface Op<T> {

        T apply() throws IOException;
    }

    @CheckForNull
    @Override
    public final FsNode node(BitField<FsAccessOption> options, FsNodeName name) throws IOException {
        return apply(() -> controller.node(options, name));
    }

    @Override
    public final void checkAccess(BitField<FsAccessOption> options, FsNodeName name, BitField<Entry.Access> types) throws IOException {
        apply(() -> {
            controller.checkAccess(options, name, types);
            return null;
        });
    }

    @Override
    public final void setReadOnly(BitField<FsAccessOption> options, FsNodeName name) throws IOException {
        apply(() -> {
            controller.setReadOnly(options, name);
            return null;
        });
    }

    @Override
    public final boolean setTime(BitField<FsAccessOption> options, FsNodeName name, Map<Entry.Access, Long> times) throws IOException {
        return apply(() -> controller.setTime(options, name, times));
    }

    @Override
    public final boolean setTime(BitField<FsAccessOption> options, FsNodeName name, BitField<Entry.Access> types, long value) throws IOException {
        return apply(() -> controller.setTime(options, name, types, value));
    }

    @Override
    public final InputSocket<? extends Entry> input(BitField<FsAccessOption> options, FsNodeName name) {
        return new Input(controller.input(options, name));
    }

    private final class Input extends AbstractInputSocket<Entry> {

        private final InputSocket<? extends Entry> socket;

        Input(final InputSocket<? extends Entry> socket) {
            this.socket = socket;
        }

        @Override
        public Entry target() throws IOException {
            return apply(socket::target);
        }

        @Override
        public InputStream stream(OutputSocket<? extends Entry> peer) throws IOException {
            return apply(() -> socket.stream(peer));
        }

        @Override
        public SeekableByteChannel channel(OutputSocket<? extends Entry> peer) throws IOException {
            return apply(() -> socket.channel(peer));
        }
    }

    @Override
    public final OutputSocket<? extends Entry> output(BitField<FsAccessOption> options, FsNodeName name, @CheckForNull Entry template) {
        return new Output(controller.output(options, name, template));
    }

    private final class Output extends AbstractOutputSocket<Entry> {

        private final OutputSocket<? extends Entry> socket;

        Output(final OutputSocket<? extends Entry> socket) {
            this.socket = socket;
        }

        @Override
        public Entry target() throws IOException {
            return apply(socket::target);
        }

        @Override
        public OutputStream stream(InputSocket<? extends Entry> peer) throws IOException {
            return apply(() -> socket.stream(peer));
        }

        @Override
        public SeekableByteChannel channel(InputSocket<? extends Entry> peer) throws IOException {
            return apply(() -> socket.channel(peer));
        }
    }

    @Override
    public final void make(BitField<FsAccessOption> options, FsNodeName name, Entry.Type type, @CheckForNull Entry template) throws IOException {
        apply(() -> {
            controller.make(options, name, type, template);
            return null;
        });
    }

    @Override
    public final void unlink(BitField<FsAccessOption> options, FsNodeName name) throws IOException {
        apply(() -> {
            controller.unlink(options, name);
            return null;
        });
    }
}
